SpockOpaleye/src/ApiBase.hs

```diff
@@ -0,0 +1,47 @@
+{-# LANGUAGE FlexibleContexts       #-}
+{-# LANGUAGE FlexibleInstances      #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE MultiParamTypeClasses  #-}
+{-# LANGUAGE OverloadedStrings      #-}
+{-# LANGUAGE TemplateHaskell        #-}
+
+module ApiBase where
+
+import           Control.Lens
+import qualified Data.Profunctor.Product.Default as D
+import           Data.Time                       (UTCTime, getCurrentTime)
+import           Database.PostgreSQL.Simple
+import           DataTypes
+import           Opaleye
+import           OpaleyeDef
+import           OpaleyeTypes
+import           Prelude                         hiding (id)
+
+create_row ::(
+    HasCreatedat columnsW (Maybe (Column PGTimestamptz)),
+    HasUpdatedat columnsW (Column PGTimestamptz),
+    D.Default Constant incoming columnsW, D.Default QueryRunner returned row)
+    => Connection -> Table columnsW returned -> incoming -> IO row
+create_row conn table item = do
+  current_time <- fmap pgUTCTime getCurrentTime
+  let itemPg = (constant item) & createdat .~ (Just current_time) & updatedat .~ (current_time)
+  fmap head $ runInsertManyReturning conn table [itemPg] (\x -> x)
+
+update_row :: (
+    HasUpdatedat haskells UTCTime
+    , D.Default Constant haskells columnsW
+    , D.Default Constant item_id (Column PGInt4)
+    , HasId haskells item_id
+    , HasId columnsR (Column PGInt4)
+    )
+    => Connection -> Table columnsW columnsR -> item_id -> haskells -> IO haskells
+update_row conn table it_id item = do
```

Comment:
> What we are trying to prevent is ending up with inconsistent data. I gave you an example earlier.

I meant an actual piece of code that you would genuinely write this way:

* Fetch a record from the DB (and the record contains an ID field)
* Change the values, but not the ID to obtain a new record (because of immutability and all that)
* Write the record back to the DB, but to overwrite a **different row**